x/gov/keeper/proposal.go

```diff
@@ -95,6 +95,38 @@ func (keeper Keeper) SubmitProposal(ctx sdk.Context, messages []sdk.Msg, metadat
 	return proposal, nil
 }
 
+// CancelProposal will cancel proposal before the voting period ends
+func (keeper Keeper) CancelProposal(ctx sdk.Context, proposalID uint64, proposer string) error {
+	proposal, ok := keeper.GetProposal(ctx, proposalID)
+	if !ok {
+		return sdkerrors.Wrapf(types.ErrProposalNotFound, "proposal is not found with %d id", proposalID)
+	}
+
+	// Check creator of the proposal
+	if proposal.Proposer != proposer {
+		return sdkerrors.Wrapf(types.ErrInvalidProposer, "invalid proposer %s", proposer)
+	}
+
+	// Check if proposal is active or not
+	if (proposal.Status != v1.StatusDepositPeriod) && (proposal.Status != v1.StatusVotingPeriod) {
+		return sdkerrors.Wrapf(types.ErrInactiveProposal, "%d", proposalID)
+	}
+
+	// Check proposal voting period is ended.
+	if proposal.VotingEndTime != nil && proposal.VotingEndTime.Before(ctx.BlockTime()) {
```

Comment:
is there a case where proposal.VotingEndTime can be nil?

Reply:
Yes, if the proposal still in deposit period

Reply:
Let's add that note into the comment.